sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateStoreSuite.scala

```diff
@@ -1718,6 +1845,240 @@ abstract class StateStoreSuiteBase[ProviderClass <: StateStoreProvider]
+    )
+
+    val conf = new SparkConf().setMaster("local").setAppName("test")
+
+    withSpark(SparkContext.getOrCreate(conf)) { sc =>
+      withCoordinatorRef(sc) { coordinatorRef =>
+        val rootLocation = s"${Utils.createTempDir().getAbsolutePath}/spark-51596-signaling"
+        val providerId =
+          StateStoreProviderId(StateStoreId(rootLocation, 0, 0), UUID.randomUUID)
+
+        // Create the first provider - this loads it into StateStore's loadedProviders map
+        StateStore.get(
+          providerId,
+          keySchema, valueSchema, NoPrefixKeyStateEncoderSpec(keySchema),
+          0, None, None, useColumnFamilies = false, new StateStoreConf(sqlConf), new Configuration()
+        )
+
+        // Manually trigger maintenance to simulate scheduled maintenance
+        // This will start maintenance and block in our custom provider's doMaintenance()
+        val maintenanceMethod = PrivateMethod[Unit](Symbol("doMaintenance"))
+        StateStore invokePrivate maintenanceMethod()
+
+        // Wait for maintenance to actually start before continuing
+        eventually(timeout(5.seconds)) {
+          assert(SignalingStateStoreProvider.maintenanceStarted)
+          // Verify the provider is still loaded at this point
+          assert(StateStore.isLoaded(providerId))
+        }
+
+        // Create a separate thread that simulates a task thread trying
+        // to unload a provider that's currently under maintenance
+        val taskThread = new Thread("task-simulation-thread") {
+          override def run(): Unit = {
+            // Tell the coordinator this instance is now active elsewhere
+            // This makes the local instance "stale"
+            coordinatorRef.reportActiveInstance(providerId, "otherhost", "otherexec", Seq.empty)
+
+            // Loading another provider while the first is stale should trigger unloading
+            // of the stale provider through maintenance submitted by the task thread
+            val providerId2 =
+              StateStoreProviderId(StateStoreId(rootLocation, 0, 1), UUID.randomUUID)
+            StateStore.get(
+              providerId2,
+              keySchema, valueSchema, NoPrefixKeyStateEncoderSpec(keySchema),
+              0, None, None, useColumnFamilies = false, new StateStoreConf(sqlConf),
+              new Configuration()
+            )
+
+            // Signal that we've completed the operations that should trigger maintenance
+            SignalingStateStoreProvider.taskSubmittedMaintenance = true
+          }
+        }
+
+        // Start the task thread
+        taskThread.start()
+
+        // Wait for the task to execute the operations that should trigger maintenance
+        eventually(timeout(5.seconds)) {
+          assert(SignalingStateStoreProvider.taskSubmittedMaintenance)
+        }
+
+        // Now unblock the first maintenance operation
+        // This should allow the task's submitted maintenance to proceed
```

Comment:
how do you know that the submited work actually waited for the ongoing maintenance and didn't happen concurrently? I don't see that verification here